lib/SILOptimizer/Mandatory/TFCanonicalizeCFG.cpp

```diff
@@ -233,9 +257,358 @@ SESERegionBuilder::processAcyclicRegionExcludingEnd(SILBasicBlock *startBB,
+    edgesToFix.emplace_back(headerPred, header);
+  }
+
+  SILBuilder builder(header);
+  ASTContext &context = builder.getASTContext();
+
+  // Create a new header for the loop.
+  //
+  SILBasicBlock *newHeader = currentFn->createBasicBlock();
+  loop->addBasicBlockToLoop(newHeader, LI.getBase()); // Should be done first.
+  // Clone arguments and change all uses to the new header's arguments.
+  newHeader->cloneArgumentList(header);
+  SmallVector<SILValue, 8> headerArgs;
+  for (auto i : indices(header->getArguments())) {
+    header->getArgument(i)->replaceAllUsesWith(newHeader->getArgument(i));
+    headerArgs.push_back(newHeader->getArgument(i));
+  }
+  size_t oldHeaderNumArgs = header->getNumArguments();
+  header->dropAllArguments();
+  // Add phi arguments in the new header corresponding to the escaping values.
+  for (const auto &escapingValue : escapingValues) {
+    SILValue newValue = newHeader->createPHIArgument(
+        escapingValue->getType(), escapingValue.getOwnershipKind());
+    escapingValue->replaceAllUsesWith(newValue);
+  }
+  // An integer to identify the exit edge.
+  SILValue exitIndexArg = newHeader->createPHIArgument(
+      convertElementTypeToTensorValueType(
+          SILType::getBuiltinIntegerType(32, context)),
+      ValueOwnershipKind::Owned);
+  // A boolean corresponding to the stayInLoop flag.
+  newHeader->createPHIArgument(convertElementTypeToTensorValueType(
+                                   SILType::getBuiltinIntegerType(1, context)),
+                               ValueOwnershipKind::Trivial);
+
+  // Create a new latch block.
+  SILBasicBlock *latchBlock = currentFn->createBasicBlock();
+  latchBlock->cloneArgumentList(newHeader);
+  {
+    builder.setInsertionPoint(latchBlock);
+    SmallVector<SILValue, 8> latchArgs;
+    for (const SILArgument *latchArg : latchBlock->getArguments()) {
+      latchArgs.push_back(latchArg);
+    }
+    builder.createBranch(
+        getUserSourceLocation(header->getTerminator()->getDebugLocation()),
+        newHeader, latchArgs);
+  }
+
+  llvm::DenseMap<SILBasicBlock *, intmax_t> exitIndices;
+  // As a simplification, simply 0 as exit index for edges to the header
+  // even though header is not an exit block.
+  exitIndices.insert({header, 0});
+  // Identify the exit from the header and also set it to zero index.
+  SILBasicBlock* headerExit = nullptr;
+  for (SILBasicBlock* succ : header->getSuccessorBlocks()) {
+    if (loop->contains(succ))  continue;
+    assert(headerExit == nullptr && "Loop header has more than one exit node.");
+    headerExit = succ;
+  }
+  if (headerExit != nullptr) {
+    exitIndices.insert({headerExit, 0});
+  }
+  unsigned nextExitIndex = 1;
+  for (const auto &edge : edgesToFix) {
+    SILBasicBlock *src = const_cast<SILBasicBlock *>(edge.first);
+    SILBasicBlock *tgt = const_cast<SILBasicBlock *>(edge.second);
+    SILBasicBlock *newTgt = loop->contains(src) ? latchBlock : newHeader;
```

Comment:
this stmt is not obvious -- consider adding a comment. more generally provide context on what `edgesToFix` mean and why we use that.

Reply:
Added comments describing the meaning of edgesToFix. I also changed loop->contains(src) to  (src == preheader), which is easier to read and understand.